pkg/operator/staticpod/controller/guard/guard_controller.go

```diff
@@ -0,0 +1,255 @@
+	policyclientv1 "k8s.io/client-go/kubernetes/typed/policy/v1"
+	corelisterv1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/staticpod/controller/guard/bindata"
+)
+
+// GuardController is a controller that watches amount of static pods on master nodes and
+// renders guard pods with a pdb to keep maxUnavailable to be at most 1
+type GuardController struct {
+	targetNamespace, podResourcePrefix string
+	operatorName                       string
+	readyzPort                         string
+
+	nodeLister corelisterv1.NodeLister
+	podGetter  corev1client.PodsGetter
+	pdbGetter  policyclientv1.PodDisruptionBudgetsGetter
+
+	// installerPodImageFn returns the image name for the installer pod
+	installerPodImageFn   func() string
+	createConditionalFunc resourceapply.ConditionalFunction
+	deleteConditionalFunc resourceapply.ConditionalFunction
+}
+
+func NewGuardController(
+	targetNamespace, podResourcePrefix string,
+	operatorName string,
+	readyzPort string,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	kubeInformersClusterScoped informers.SharedInformerFactory,
+	podGetter corev1client.PodsGetter,
+	pdbGetter policyclientv1.PodDisruptionBudgetsGetter,
+	eventRecorder events.Recorder,
+	createConditionalFunc resourceapply.ConditionalFunction,
+	deleteConditionalFunc resourceapply.ConditionalFunction,
+) factory.Controller {
+	c := &GuardController{
+		targetNamespace:       targetNamespace,
+		podResourcePrefix:     podResourcePrefix,
+		operatorName:          operatorName,
+		readyzPort:            readyzPort,
+		nodeLister:            kubeInformersClusterScoped.Core().V1().Nodes().Lister(),
+		podGetter:             podGetter,
+		pdbGetter:             pdbGetter,
+		installerPodImageFn:   getInstallerPodImageFromEnv,
+		createConditionalFunc: createConditionalFunc,
+		deleteConditionalFunc: deleteConditionalFunc,
+	}
+
+	return factory.New().WithInformers(
+		kubeInformersForTargetNamespace.Core().V1().Pods().Informer(),
+		kubeInformersForTargetNamespace.Core().V1().Nodes().Informer(),
+	).WithSync(c.sync).ToController("GuardController", eventRecorder)
+}
+
+func getInstallerPodImageFromEnv() string {
+	return os.Getenv("OPERATOR_IMAGE")
+}
+
+func getGuardPodName(prefix, nodeName string) string {
+	return fmt.Sprintf("%s-guard-%s", prefix, nodeName)
+}
+
+func getGuardPDBName(prefix string) string {
+	return fmt.Sprintf("%s-guard-pdb", prefix)
+}
+
+func (c *GuardController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
sync methods need to be zero-API call in the no-action case.  This requires lister usage for PDBs, nodes, and pods for instance.

Reply:
Can you more clarify on what falls under the no-action case? Do you mean in order to decide if it is the no-action case the code needs to call c.createConditionalFunc/c.deleteConditionalFunc?

Reply:
> Can you more clarify on what falls under the no-action case? Do you mean in order to decide if it is the no-action case the code needs to call c.createConditionalFunc/c.deleteConditionalFunc?

in the delete case, you issue a delete because you do not use a lister first to see if there is work to do.

Reply:
I see what you mean. Both PDB cases updated. Both Apply/Delete PDB are preceded by checking the the PDB needs an update/exists.